### PR TITLE
change(nu5): disconnect from testnet peers using the first NU5 testnet rules

### PR DESCRIPTION
**Pull request by teor2345**
## Merging

This PR will fail the network tests until testnet has enough `zcashd` or Zebra instances on protocol version 170,040 or later:
- `zcashd` 4.6.0-2,
- `zebrad` 1.0.0-beta.8

## Motivation

`zcashd` is disconnecting from testnet nodes on network protocol versions lower than 170,040:
https://github.com/zcash/zips/pull/600/files#diff-b424a91f01c7799201470397654fcf582c045c318b0f91e1e15af4b7fdfc8f78R159-R160

So Zebra should do the same thing.

### Designs

Zebra has a minimum network protocol version constant, but it's currently using the `NetworkUpgrade` type.

## Solution

- Change the minimum network protocol version constant to use the network `Version` type
- Make the minimum testnet network protocol version 170,040

## Review

Anyone can review this PR.

This is based on PR #3799.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

## Follow Up Work

We have also updated the DNS seeders, we might as well update them straight to version 170,100:
- #3458 

(There is no minimum protocol version for the DNS seeders.)

### CODE REVIEW

**Review by conradoplg — DISMISSED** [review]
Looks good, thanks! Do you prefer merging this into `zcash_script_v5_api` or waiting for that to merge first? (I'm fine with both)

**Comment by teor2345** [comment]
> Looks good, thanks! Do you prefer merging this into `zcash_script_v5_api` or waiting for that to merge first? (I'm fine with both)

I'd prefer this as a separate PR, we might want to wait to merge until after the rolled back testnet is working.
(Zebra doesn't need this change as much as `zcashd` does.)

File: zebra-network/src/constants.rs
Comment (teor2345) [inline zebra-network/src/constants.rs:297]:
```suggestion
        // This is the `zcashd` 4.7.0 network protocol version,
        // after updating to the second NU5 testnet activation consensus rules,
        // but before the second NU5 testnet activation.
```

**Comment by teor2345** [comment]
@mergifyio refresh

**Comment by teor2345** [comment]
I need to check how many testnet peers are on version 170,040 before I merge this

**Review by daira — APPROVED** [review]
Looks good to me.

> We have also updated the DNS seeders, we might as well update them straight to version 170,100: #3458

~~The seeders won't be implementing the mainnet consensus rules yet, so they should stay on 170050 for now.~~ Actually I reviewed the seeder code and it doesn't appear to support anything that would change with NU5 mainnet. Still not sure what I think about this.

(Here's where the seeder protocol version is set:
https://github.com/ZcashFoundation/dnsseeder/blob/e9a2d22878d0ffe7a604871dee6fe50b807099ef/zcash/client.go#L43 .)

**Comment by teor2345** [comment]
> The seeders won't be implementing the mainnet consensus rules yet, so they should stay on 170050 for now. 

We're not expecting any network protocol changes before 170100.

> Also, #3458 was closed by a PR that doesn't do that, I think.

The version bump of the dependency in that PR increments the network version to 170100.

**Comment by teor2345** [comment]
> Still not sure what I think about this.

Fair enough, we just wanted to skip an hour or two of change/deployment work. I'll make sure we review these kinds of schedule/shortcut decisions more carefully next time.

**Comment by teor2345** [comment]
> I need to check how many testnet peers are on version 170,040 before I merge this

Let's merge after NU5 has re-activated on testnet, there seem to be at least 8 active testnet peers.

**Review by conradoplg — APPROVED** [review]
Looks good. Currently testnet has 14 peers on 170050 and 18 on 170015 so it seems OK to go ahead

image

After switching to this PR:

image

**Comment by teor2345** [comment]
> Looks good. Currently testnet has 14 peers on 170050 and 18 on 170015 so it seems OK to go ahead

Thanks for doing these detailed checks, I didn't quite get to them yesterday!

**Pull request merged**
